### DISCUSSION TITLE
Make media_source an entity component, add MediaSourceEntity base class

### DISCUSSION THREAD

**Discussion by chishm** [home-assistant/architecture#675] [Entity Models] [answered]
# Proposal
I'd like to make `media_source` into an entity component and add a `MediaSourceEntity` base class for deriving media sources from. Each MediaSourceEntity would be visible at the root level of the media browser, like "Local Media" and "Xbox Game Media" (for example) currently are.

The object ID (the 2nd part of the `entity_id`) of the entity would be used for its name in `media_source://` URIs. The `entity_picture` would be used for the thumbnail when browsing at the root level. The entity `name` would be displayed as the source name. And any entities that are not `available` would not be displayed. Being available indicates that the media source (e.g. file server) is online and browsable. Entities would be categorised as "diagnostic" so they don't appear in the default Lovelace view.

This would be an extension of the existing approach, where currently each platform is responsible for managing its own contents and is only exposed as one "entity" at the top-level. Existing sources, currently derived from `MediaSource`, would still be supported and won't need to be turned into entities (although migration is fully welcome).

# Benefits
Integrations that can provide multiple independent media sources currently have to manage those sources within the individual integration. This leads to duplicate code (lots of `async_parse_identifier` with slight differences), or just no support for more than one media source (e.g. the `jellyfin` component).

Managing sources as Entity objects allows them to be reconfigured like any other entity (assuming a `unique_id`), so they can be renamed, re-iconed, or disabled completely.

# Not in this proposal
I am not proposing any services for `MediaSourceEntity`s here. The sources currently provide a back-end service in the form of `async_browse_media` and `async_resolve_media` that are not directly callable via service calls, and this state of affairs would be maintained.

I am not proposing any `state_attributes` for `MediaSourceEntity`. There may be some that are common and useful, like "newest added item" or "total items", but I'm keeping this proposal to the basics.

# Why?
I'm proposing this because it was the best approach I found while working on adding a new `media_source` platform, in this case `dlna_dms`. I have a [WIP branch](https://github.com/chishm/core/tree/dlna-dms) that already has a commit to implement [`MediaSourceEntity`](https://github.com/chishm/core/commit/cf69a985c11768d0d01d3939ca0b36c56967b0ce).

# Work to do
I've already written the code to implement this, and I will also do the following if this proposal is accepted:
* Unit tests to cover the entity component part of `media_source`
* Developer documentation for `MediaSourceEntity`

Future work, if desired:
* Migrate `media_source.local_source` to be based on `MediaSourceEntity`.
* Migrate `jellyfin` to `MediaSourceEntity` and allow multiple config entries (for multiple Jellyfin servers).
* Browse a `MediaSourceEntity` as its default action when clicked on in the front end.

**Comment by allenporter** [comment]
One topic that came up in the nest media player review was.adding permissions checks for downloading media.

My impression is that user permissions are not yet a UI feature, but that checks are being proactively put in place.

In the nest pr it was added to the API endpoint where media is downloaded. (Aside: The API is in terms of devices, however the permission check is in terms of entities.)

We were talking on the PR about also adding this check in the media source itself also, but it didn't seem possible with the current interface as far as I could see. Perhaps a media source entity as proposed here could move that permission check in the media source APIs.

**Comment by balloob** [comment, marked as answer]
I don't like the idea of adding an entity purely for organizational purposes of a single UI. The media source UI already allows defining pictures and names for entries, and as an integration you can completely decide what to return as your "root" without relying on entities.

**Reply by chishm** [reply]
Thank you for your reply. I hope to convince you there's more benefit to `MediaSourceEntity` than just the configuration UI.

For users, it exposes the state of the media source, e.g. "on" or "unavailable", which can be used in automations as well as viewed in the UI.

For integration implementations, it:
* defines the well-known `update()` / `async_update()` method for polling sources
* removes the hassle of creating a unique name (slug) per source
* provides an easy way to enable/disable a source, both in response to the user ("enable entity" checkbox) and in response to the source (going "unavailable" when no connection exists)
* (in future) allows per-source device/entity permission checks to be implemented consistently, as suggested by @allenporter 

I would have to re-implement all of these features internally in `dlna_dms` if sources aren't entities.

Finally, it flattens the media tree slightly, reducing the number of clicks to get to an item.

<details><summary>Example flattening</summary>

For example, it could go from:
* Local
    * Sounds
        * File 1
        * File 2
    * Videos
        * File 3
* DLNA
    * Media server
        * Music
            * Song 1
            * Song 2
        * Movies
            * Movie 1
    * Video recorder
        * Show 1

Into the flatter hierarchy:

* Sounds
    * File 1
    * File 2
* Videos
    * File 3
* Media server
    * Music
        * Song 1
        * Song 2
    * Movies
        * Movie 1
* Video recorder
    * Show 1
</details>

**Reply by allenporter** [reply]
I recently added a Nest Media Source, and the hierarchy is a top level *Nest* directory, then one directory below for each camera (e.g. 4 cameras). With this proposal, there would be one top level item for every camera (e.g. 4 top level items). I don't have a clear idea for the relationship between a camera device, its camera entity and its media source entity, so not sure if this is better or worse -- but then again, cameras are a little strange since i've heard them be referred to as virtual entities.

**Reply by balloob** [reply]
There is no reason to poll sources. With the media source it is the user that will indicate when it wants to fetch media, we shouldn't keep fetching all media files in case a request comes in.

You need a unique ID anyway because how else will you refer to the resource?

A source being unavailable is easy to spot, as it might not show up in the list or maybe it shows up but returns an error when browsed.

**Reply by chishm** [reply]
Thanks again for taking the time to consider this proposal. I hope I'll manage to sway you with my responses below, but if it's still a no-go then I'll try to rework `dlna_dms` to not rely on entities.

> There is no reason to poll sources. With the media source it is the user that will indicate when it wants to fetch media, we shouldn't keep fetching all media files in case a request comes in.

In the `dlna_dms.async_update` case, it's to check if the network connection is still available, rather than timing out x seconds after the user tries to browse. This keeps the browsing UI responsive, because a source will be immediately missing from the root browse view if it's unavailable, rather than the user getting a browse error after waiting for a timeout after clicking on a potentially missing source.

> You need a unique ID anyway because how else will you refer to the resource?

Indeed, and I meant it removes the need to write a custom ID generator function because the `entity_id` does the job quite well already. Otherwise I'd need to replicate what `EntityRegistry.async_generate_entity_id` does.

> A source being unavailable is easy to spot, as it might not show up in the list or maybe it shows up but returns an error when browsed.

Yep, and I prefer the first option (see my description of the use for `dlna_dmr.async_update` above). But that doesn't help for non-interactive cases, like automations, where currently the state of a media source is unknowable.

**Reply by allenporter** [reply]
Does it make sense to provide a link between the Devices page and associated Media Player directories for that device? I'm thinking of the case of a camera, to view its event snapshots. 

If that is useful: what is the best way to make that linkage? I could imagine an entity being that link, but also just a media player url could work.
